docs/rfcs/021-metering.md

```diff
@@ -0,0 +1,175 @@
+# Consumption tracking
+
+# Goals
+
+This proposal is made with two mostly but not entirely overlapping goals:
+
+* Collect info that is needed for consumption-based billing
+* Cross-check AWS bills
+
+
+# Metrics
+
+There are six metrics to collect:
+
+* CPU time. Wall clock seconds * the current number of cores. We have a fixed ratio of memory to cores, so the current memory size is the function of the number of cores. Measured per each `endpoint`.
+
+* Traffic. In/out traffic on the proxy. Measured per each `endpoint`.
+
+* Written size. Amount of data we write. That is different from both traffic and storage size, as only during the writing we
+
+  a) occupy some disk bandwidth on safekeepers
+  
+  b) necessarily cross AZ boundaries delivering WAL to all safekeepers
+
+  Each timeline/branch has at most one writer, so the data is collected per branch.
+
+* Synthetic storage size. That is what is exposed now with pageserver's `/v1/tenant/{}/size`. Looks like now it is per-tenant. (Side note: can we make it per branch to show as branch physical size in UI?)
+
+* Real storage size. That is the size of the tenant directory on the pageservers disk. Per-tenant.
+
+* S3 storage size. That is the size of the tenant data on S3. Per-tenant.
+
+That info should be enough to build an internal model that predicts AWS price (hence tracking `written data` and `real storage size`). As for the billing model we probably can get away with mentioning only `CPU time`, `synthetic storage size`, and `traffic` consumption.
+
+# Services participating in metrics collection
+
+## Proxy
+
+Proxy is the only place that knows about traffic flow, so it tracks it and reports it with quite a small interval, let's say 1 minute. A small interval is needed here since the proxy is stateless, and any restart will reset accumulated consumption. Also proxy should report deltas since the last report, not an absolute value of the counter. Such kind of events is easier to integrate over a period of time to get the amount of traffic during some time interval.
+
+Example event:
+
+```json
+{
+    "metric": "traffic_in_bytes",
+    "type": "increment",
+    "endpoint_id": "blazing-warrior-34",
+    "event_start_time": ...,
+    "event_stop_time": ...,
+    "value": 12345454,
+}
+```
+
+Since we report deltas over some period of time, it makes sense to include `event_start_time`/`event_stop_time` where `event_start_time` is the time of the previous report. That will allow us to identify metering gaps better (e.g., failed send/delivery).
```

Comment:
not sure if graphite and statsd are the way to go.
they force to fit into the typical format and metadata for their destinations, e.g.
- graphite would have one timestamp and some clients may be optimized for a predefined resolution, tags are a late addition to graphite, typically you would have a dot.separated.metric.name
- does statsd include the timestamp of the measurement?

there's also a question of the support on the recipient. e.g. i haven't seen billing vendors supporting graphite as an input, not sure we want to lock ourselves into the format & schema transformation on the way towards the consumers

Reply:
IMHO OTel should be more modern, developing, flexible and more widely supported for the consumption too compared to graphite, stasd and influx.

Reply:
Collection code could be fairly involved, e.g. we'll probably end up with saving some disk state about last sent values. So decent collector could be a good reason to stick to some existing format. I went checked:
* https://github.com/markstory/rust-statsd -- forces UDP, that will complicate the receiver side
* https://github.com/influxdb-rs/influxdb-rust -- assumes influxdb on the other side, so I guess receiver has emulate a bit more then a message format to pretend to be an influx instance
* https://github.com/tureus/graphite-rust -- not maintained for 7 years

Also formats there are not super attractive -- some custom text-based representations with spaces and commas.

With that I'd rather write billing metrics collection manually and own it -- we don't need a super involved logic there as well. I think of that reporting pipeline more like kafka events or RPC append calls, rather then a usual observability metrics.

> I'm not sure start time actually helps that much: if we have constant interval, we already know that the start time is now - interval.

It is a bit different with CPU time in this proposal -- it is num_cores*cpu_time, so it can report 240 seconds when using 4 cores during 60 seconds.

Reply:
> IMHO OTel should be more modern
Hence ”such as”, didn’t mean exclude any specific protocols.

> there's also a question of the support on the recipient. e.g. i haven't seen billing vendors supporting graphite as an input

So recipient would be the collector agent. What happens after that would just have more options as the selection of protocols would be larger.

Reply:
> https://github.com/markstory/rust-statsd -- forces UDP, that will complicate the receiver side

Hence suggesting to use collection agent in between. They all support udp with statsd as because it's very common to use standard and has low impact even with very high volume of metrics sent.

> https://github.com/influxdb-rs/influxdb-rust -- assumes influxdb on the other side, so I guess receiver has emulate a bit more then a message format to pretend to be an influx instance

vmagent and telegraf will happily take influxdb, although I think vector might be more interesting here, as it for example supports unit testing with configurations. 


> Also formats there are not super attractive -- some custom text-based representations with spaces and commas.

What is more custom than own format? :) But that's honestly besides the point. These are designed for very high volume metrics and therefore opt to simpler format, which takes less cpu cycles to parse. 

But as a clarification, what I'm, thinking is more like:
```


  ┌───────────┐                     ┌─────────┐
  │ pageserver├────────┐            │         │
  └───────────┘        │            │ console │
                     ┌─▼──────┐     │         │
  ┌───────────┐      │        ├────►└───┬─────┘
  │ proxy     ├──────► vector │         │
  └───────────┘      │        │     ┌───▼─────┐
                     └─▲──────┴────►│ external│
  ┌───────────┐        │        ?   │ services│
  │ autosaler ├────────┘            │         │
  └───────────┘                     └─────────┘
```

Then we would be able to send metrics to vector even ridiculously often, have it aggregate and act as a durable queue and have it either push to somewhere (even to external services, while not necessarily the best option, depending on service) or allow console to scrape e.g using prometheus style scraping.

There very likely has to be a queue in between console and metrics sent in any case, and it has to be something pretty simple and small, which won't be updated often.

Reply:
@lassizci few questions about the picture:

* what is used on pageserver/proxy side to report metrics in this picture? prometheus endpoint?
* how can console receive that stream from vector? any existing packages?
* can vector downsample the data? e.g. poll proxy each 15 sec but push to the console each five minutes.

Generally I like that idea -- it could remove some amount of complexity about pushing the events / doing retries from the services themselves. OTOH I feel a bit uneasy about introducing quite a big service in the middle of our billing pipeline (however may be it is okay since we mainly can lose traffic events but other thing can be regenerated from the data stored in services)

Reply:
@kelvich I think I would
* prefer to push with statsd to vector
* vector can for example expose prometheus format /metrics
* vector can also aggregate data. It can actually do a lot of things with [transforms](https://vector.dev/docs/reference/configuration/transforms/), [VRL](https://vector.dev/docs/reference/vrl/)

@antonyc mentioned having used it. Would you have insights about could it suit for this kind of use case?
I've just tried briefly for deduplicating rust stacktraces from plaintext logs, which actually seemed possible with VRL and combining transforms. [Unit testing](https://vector.dev/docs/reference/configuration/unit-tests/) is also possible, which could probably be considered as a requirement, considering we are talking about biling events.

> OTOH I feel a bit uneasy about introducing quite a big service in the middle of our billing pipeline

Understand the concern and it is a service on its' own and can get complicated too. On the other hand it's quite simple to get started with. A very simplified 2 minute demo would be:

```
$ cat vector.yaml 
data_dir: /tmp/vector
sources:
  statsd_source:
    type: statsd
    address: 0.0.0.0:8125
    mode: udp
transforms:
  aggregator:
    type: aggregate
    inputs:
      - statsd_source
    interval_ms: 5000
sinks:
  debugger:
    type: console
    inputs:
      - statsd_source
    target: stdout
    encoding:
      codec: json
  prom:
    type: prometheus_exporter
    inputs:
      - aggregator
    address: 0.0.0.0:9000
    default_namespace: mymetrics
    
$ ./vector -c vector.yaml
```

```
$ while [ 1 ]; do curl http://localhost:9000/metrics; sleep 1; done
```

```
$ cat metrics.py 
import socket, time

sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
sock.sendto(b"my_metric:1|c|#tenant:foo,", ("localhost", 8125))

time.sleep(6)

for i in range(1000):
  sock.sendto(b"my_metric:1|c|#tenant:foo,", ("localhost", 8125))

$ python metrics.py
```

So sending counter 1, waiting for 6 seconds, then sending 1000.

Reply:
Could also post over http from vector, but wasn’t as quick to set up

Reply:
I took a closer look on Vector. Originally I liked it a lot, so explored on how we can use and almost with each of the use cases there is some nit in details.

* Using it to export data to segment or directly to the snowflake. That is not currently supported https://github.com/vectordotdev/vector/issues/1460 and https://github.com/vectordotdev/vector/issues/4365
* With some VRL we could calculate sum of increments and publish result in redis key for console to query it. But currently key/values updates are not supported, only lists/channels.
* Using existing client library could harden our pushes. Statsd client is super simple, just UDP packets without acks. So we would even have to change approach to deltas. That feels like a downgrade compared to http. I went through the list of sources and one possible push client is open telemetry, but vector only supports otel logs, not metrics. There is also generic http source which we could use but then the pipeline is `neon service -> (http) -> vector -> (http) -> console -> segment`. And vector doesn't add a lot of value in such a setup.
* Alternative would be to actually don't do any pushes from services, let `vector` scrape prometheus endpoints and then deliver them to the console. That setup has a nice property of not changing things in our services. However it is still wasteful to scrape everything and there is no good way to filter out unchanged events in the pipeline. So IMO better to go with pushes.

So with that I'd suggest to just POST jsons to the console for now.

Also, having few more chats with @ololobus on possible future ways to avoid segment->snowflake data flow and ways to show traffic / compute usage in console on different intervals I think we will end up with Kafka/Redpanda/Materialize for that pipeline later. But still with console acting as a proxy for all events we can play with few approaches only changing one component. (cc @vadim2404 -- I'm about to give up on no Kafka policy)

Reply:
(or we could use existing VictoriaMetrics to show that info)

Reply:
Fair enough, doesn't have to be vector. Let me still make a couple points that might be worth thinking about in general:

> Using it to export data to segment or directly to the snowflake

This is the part I would actually be the most hesitant about (also hence the `? `). I think what we collect is rather standard -> use standard collection methods. But what we want to have eventually in Snowflake is much more open ended -> the last mile is probably where the magic needs to happen. Probably would prefer separate microservice from console, but that could happen later too. Scalability requirements are different for different parts of console and that will be an issue at some point probably.

> With some VRL we could calculate sum of increments and publish result in redis key for console to query it. But currently key/values updates are not supported, only lists/channels.

If redis is treated like a queue, list is probably the right data structure. If it was a key where value is updated constantly, it would be essentially the same as scrape.

 > So we would even have to change approach to deltas.
  
I think there are couple mixed things here: my point relates to @stepashka 's [comment](https://github.com/neondatabase/neon/pull/2884#discussion_r1029098828) a little. 

* We don't have reliable absolute counters. Proxy [increments counters](https://github.com/neondatabase/neon/blob/05db6458df5430fbd659962d8b6c5a16ef735da9/proxy/src/proxy.rs#L276-L286) on send/receive. The internal registry is cleared when proxy is restarted.
* But if we could receive every counter increment and would be able to aggregate these as close as possible to the source (like something receiving on every node), and then have a reliable method to pass these forward, we would be able to worry less about loss of signal. It's then also separate concern whether Segment/Snowflake wants deltas or absolute counter.

>  That feels like a downgrade compared to http
* proxy doing cross continent http might have concerns as the latencies are large, which is why this would be nice to be able to offload from proxy to _something_ that's local, either pushing to console or to queue where console can read
* I wouldn't use statsd either, unless the aggregator was something that supports it out of the box. [Datadog's format](https://docs.datadoghq.com/developers/dogstatsd/datagram_shell/?tab=metrics#datagram-format) is rather de-facto nowdays and that's what vector uses too. It's not end to end format, but neither is prometheus.

Reply:
I think we should start with something that would allow us to move quickly, to give the needed agility to understand what we bill for, how frequently we count usage, and bill for it. That would allow us to strongly define business logic for billing and understand the limits of our solution to improve the bottlenecks later. 
A pipeline with vector accepting events via HTTP (any protocol it supports really) + Kafka as a bus + Kafka-Snowflake connector to store raw events in Snowflake + Kafka to Postgres connector to store events in our own single RDS could be a reliable solution. And the bills/usage could be calculated in a separate service later. That would deliver the final numbers in terms of minutes to the customers. But it's also a very expensive pipeline (did not do a napkin maths, but it seems to take very much time to build correctly). So we probably need to iterate?

Reply:
> Pushing everything to console at some point will create scalability problems

Can you share your though process behind that? My back of the napkin math tells me that hardly ever can burn one cpu core. May be we just have somehow different models of how this thing should operate?

Reply:
> I think it should say control-plane, not console

Currently it is console. If we plan to implement it in console then it is console and not control plane. 

> Can you share your though process behind that?

I think the main problem with that is that shape of our infra is underspecified in terms of how many tenants we can handle per "cell" (what cell actually is?). I imagine that we can have hundred pageservers in one cell (taking <100 as up to 100 from @ololobus comment above). One pageserver can have thousands (tens of thousands?) of assigned tenants. Do we have statistics on how many of them can be active at the same time? We can have multiple active computes per tenant (several timelines). For one entity metric is reported from 6 places one of them being the console itself. Other ones are 3 safekeepers, proxy and pageserver. Lets assume we have 10k tenants on each pageserver and 20% of them are active (not sure how realistic is it). Then we have 2k of active tenants per pageserver which is 2k*100 = 200k. Then we have safekeepers and proxy, then we have multiple timelines and multiple time series per each one. We should do batching, because doing one request per tenant will be quite inefficient. So we have 2k*100ps tenants, we have 3x on safekeepers and one for proxy. Assuming one time series per tenant it is 600k data points per minute. Obviously it wont directly translate to RPS. How do we do batching? Even if we consolidate all data from each node in one request then requests will be big. With these numbers it looks necessary to do batching on the client side. If it is our own protocol then we need to code and maintain this. If we use stable third party protocol it may be way easier to support. Here we can split transport from schema and discuss it separately

> I think that current API backends can ingest thousands of requests per second.

What kind of processing they do? It is simple to do many RPS if there is not much to do (pipe request to /deb/null) but sending data over the network to another system is different.

Were there any load tests for API backends?

> My back of the napkin math tells me that hardly ever can burn one cpu core. 

What was your path to "hardly ever can burn one cpu core"?

Looks like design of this pipeline will be with us for a long time so to me it makes sense to go beyond napkin math and plan it carefully with load testing involved and various tools tested against expected volume of data

Reply:
I think in terms of cell capacity, the upper bound limit is dictated by the number of pods/nodes the k8s cluster could handle, and the most recent figures k8s has are 5000 nodes, 150000 pods per cluster. These are recent figures for bleeding edge deployments and k8s v1.25, I doubt these figures are easy to have out of the box in a managed AWS Kubernetes. In the worst case, that would mean 150000 timelines in a cell, and 150000 tenants in a cell. Which could mean 150k*6=900k events per 1 minute if we send them each minute, which is 15k RPS load. This is rather big volume, and it's too much for the services we intend to use as the consumers of usage data and invoice producers. But we need to get to that volume first.